File: flutter/lib/src/components/checkbox.dart
import 'package:lemonade_design_system/lemonade_design_system.dart';

/// Represents the state of a checkbox.
enum CheckboxStatus {
  /// The checkbox is checked.
  checked,

  /// The checkbox is unchecked.
  unchecked,

  /// The checkbox is in an indeterminate state.
  indeterminate,
}

/// {@template LemonadeCheckbox}
/// A form control that lets users select one or more options from a set.
///
/// A [LemonadeCheckbox] displays a square checkbox with an optional label
/// and support text. Supports checked, unchecked, and indeterminate states
/// for flexible selection logic.
///
/// ## Example
/// ```dart
/// LemonadeCheckbox(
///   status: CheckboxStatus.checked,
///   onChanged: () {
///     setState(() {
///       status = status == CheckboxStatus.checked
///           ? CheckboxStatus.unchecked
///           : CheckboxStatus.checked;
///     });
///   },
///   label: 'Option 1',
///   supportText: 'Additional information about this option',
/// )
/// ```
///
/// See also:
/// - [LemonadeRadioButton], for single selection scenarios
/// - [LemonadeTheme], which provides the design tokens
/// {@endtemplate}
class LemonadeCheckbox extends StatelessWidget {
  /// {@macro LemonadeCheckbox}
  const LemonadeCheckbox({
    required this.status,
    required this.onChanged,
    this.label,
    super.key,
    this.supportText,
    this.enabled = true,
    this.semanticIdentifier,
    this.semanticLabel,
  });

  /// {@template LemonadeCheckbox.status}
  /// The current status of the checkbox.
  ///
  /// Can be [CheckboxStatus.checked], [CheckboxStatus.unchecked],
  /// or [CheckboxStatus.indeterminate].
  /// {@endtemplate}
  final CheckboxStatus status;

  /// {@template LemonadeCheckbox.onChanged}
  /// A callback that is invoked when the user clicks the checkbox.
  /// {@endtemplate}
  final VoidCallback onChanged;

  /// {@template LemonadeCheckbox.label}
  /// The primary text label displayed next to the checkbox.
  /// {@endtemplate}
  final String? label;

  /// {@template LemonadeCheckbox.supportText}
  /// Optional secondary text displayed below the label.
  ///
  /// Use this to provide additional context or information about the option.
  /// If null, no support text is shown.
  /// {@endtemplate}
  final String? supportText;

  /// {@template LemonadeCheckbox.enabled}
  /// Whether the checkbox is interactive.
  ///
  /// When false, the control is displayed with reduced opacity and does not
  /// respond to taps. Defaults to true.
  /// {@endtemplate}
  final bool enabled;

  /// {@template LemonadeCheckbox.semanticIdentifier}
  /// An identifier for the checkbox used for accessibility and testing.
  /// {@endtemplate}
  final String? semanticIdentifier;

  /// {@template LemonadeCheckbox.semanticLabel}
  /// A label for the checkbox used for accessibility purposes.
  ///
  /// If not provided, the [label] text is used for accessibility.
  /// {@endtemplate}
  final String? semanticLabel;

  @override
  Widget build(BuildContext context) {
    final theme = LemonadeTheme.of(context);
    final spaces = theme.spaces;
    final colors = theme.colors;
    final opacity = theme.opacity;

    return Semantics(
      identifier: semanticIdentifier,
      label: semanticLabel ?? label,
      checked: status == CheckboxStatus.checked,
      enabled: enabled,
      child: GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: enabled ? onChanged : null,
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Container(
              width: theme.sizes.size550,
              height: theme.sizes.size600,
              alignment: Alignment.center,
              child: _CheckboxControl(
                status: status,
                enabled: enabled,
              ),
            ),
            if (label != null) ...[
              SizedBox(width: spaces.spacing200),
              Flexible(
                child: Opacity(
                  opacity: enabled ? 1.0 : opacity.state.opacityDisabled,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    mainAxisSize: MainAxisSize.min,
                    children: <Widget>[
                      Text(
                        label!,
                        style: theme.typography.bodyMediumMedium.apply(
                          color: colors.content.contentPrimary,
                        ),
                      ),
                      if (supportText != null) ...[
                        Text(
                          supportText!,
                          style: theme.typography.bodySmallRegular.apply(
                            color: colors.content.contentSecondary,
                          ),
                        ),
                      ],
                    ],
                  ),
                ),
              ),
            ],
          ],
        ),
      ),
    );
  }
}

/// Internal checkbox control widget that renders the square selection indicator
class _CheckboxControl extends StatelessWidget {
  const _CheckboxControl({
    required this.status,
    required this.enabled,
  });

  final CheckboxStatus status;
  final bool enabled;

  @override
  Widget build(BuildContext context) {
    final theme = LemonadeTheme.of(context);
    final colors = theme.colors;

    // Resolve colors based on state
    final Color backgroundColor;
    final Color borderColor;
    final Color iconColor;

    final size = theme.sizes.size550;
    final borderWidth = theme.border.base.border50;
    final borderRadius = theme.radius.radius150;

    if (!enabled) {
      backgroundColor = status == CheckboxStatus.unchecked
          ? colors.background.bgElevatedHigh
          : colors.background.bgElevatedHigh;
      borderColor = status == CheckboxStatus.unchecked
          ? colors.border.borderNeutralMedium
          : const Color(0x00000000);
      iconColor = colors.content.contentTertiary;
    } else if (status == CheckboxStatus.unchecked) {
      backgroundColor = colors.background.bgDefault;
      borderColor = colors.border.borderNeutralHigh;
      iconColor = colors.content.contentPrimaryInverse;
    } else {
      backgroundColor = colors.background.bgBrandHigh;
      borderColor = const Color(0x00000000);
      iconColor = colors.content.contentPrimaryInverse;
    }

    final showBorder = status == CheckboxStatus.unchecked || !enabled;

    return SizedBox(
      width: size,
      height: size,
      child: DecoratedBox(
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(borderRadius),
          color: backgroundColor,
          border: showBorder
              ? Border.all(
                  color: borderColor,
                  width: borderWidth,
                )
              : null,
        ),
        child: AnimatedSwitcher(
          duration: const Duration(milliseconds: 150),
          switchInCurve: Curves.easeInOut,
          switchOutCurve: Curves.easeInOut,
          transitionBuilder: (child, animation) {
            return FadeTransition(
              opacity: animation,
              child: ScaleTransition(
                scale: animation,
                child: child,
              ),
            );
          },
          child: status != CheckboxStatus.unchecked
              ? Center(
                  key: ValueKey(status),
                  child: LemonadeIcon(
                    icon: status == CheckboxStatus.checked
                        ? LemonadeIcons.checkSmall
                        : LemonadeIcons.minus,
                    color: iconColor,
                  ),
                )
              : const SizedBox.shrink(
                  key: ValueKey('empty'),
                ),
        ),
      ),
    );
  }
}

File: kmp/ui/src/commonMain/kotlin/com/teya/lemonade/Checkbox.kt
package com.teya.lemonade

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.focusable
import androidx.compose.foundation.hoverable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.interaction.collectIsHoveredAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.teya.lemonade.core.CheckboxStatus
import com.teya.lemonade.core.LemonadeAssetSize
import com.teya.lemonade.core.LemonadeIcons
import com.teya.lemonade.core.LemonadeTextStyle

/**
 * A form control that lets users select one or more options from a set.
 *  Supports checked, unchecked, and indeterminate states for flexible selection logic.
 *
 * ## Usage
 * ```kotlin
 * LemonadeUi.Checkbox(
 *     status = CheckboxStatus.Checked,
 *     onCheckboxClicked = { /* on box toggle behaviour */ },
 *     label = "Label",
 *     supportText = "Support Text",
 *     enabled = false,
 * )
 * ```
 *
 * @param status The current [CheckboxStatus] of the checkbox.
 * @param onCheckboxClicked A lambda that is invoked when the user clicks the checkbox.
 * @param label The primary text label displayed next to the checkbox.
 * @param modifier The [Modifier] to be applied to the entire component.
 * @param interactionSource Optional [MutableInteractionSource] used to observe interaction states
 *  like hover and press to drive visual feedback.
 * @param supportText Optional secondary text displayed below the label. If null, it is not shown.
 * @param enabled A boolean that controls the enabled state of the checkbox.
 */
@Composable
public fun LemonadeUi.Checkbox(
    status: CheckboxStatus,
    onCheckboxClicked: () -> Unit,
    label: String,
    modifier: Modifier = Modifier,
    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
    supportText: String? = null,
    enabled: Boolean = true,
) {
    val props = platformCheckboxPropertiesProperties.copy(focusVisible = true)
    Row(
        horizontalArrangement = Arrangement.spacedBy(space = LocalSpaces.current.spacing200),
        modifier = modifier
            .clickable(
                role = Role.Checkbox,
                enabled = enabled,
                onClick = onCheckboxClicked,
                interactionSource = interactionSource,
                indication = null,
            )
            .hoverable(interactionSource = interactionSource),
    ) {
        CoreCheckbox(
            status = status,
            interactionSource = interactionSource,
            onCheckboxClicked = onCheckboxClicked,
            enabled = enabled,
            checkboxPlatformProps = props,
        )

        Column(
            modifier = if (enabled) {
                Modifier
            } else {
                Modifier.alpha(alpha = LocalOpacities.current.state.opacityDisabled)
            },
        ) {
            LemonadeUi.Text(
                text = label,
                textStyle = props.labelStyle,
            )

            if (supportText != null) {
                LemonadeUi.Text(
                    text = supportText,
                    textStyle = props.supportTextStyle,
                    color = LocalColors.current.content.contentSecondary,
                )
            }
        }
    }
}

/**
 * This composable displays only the visual checkbox element. It's useful for custom layouts
 * where the label is handled separately.
 *
 * ## Usage
 * ```kotlin
 * LemonadeUi.Checkbox(
 *     status = CheckboxStatus.Indeterminate,
 *     onCheckboxClicked = { /* on box toggle behaviour */ },
 *     enabled = false,
 * )
 * ```
 *
 * @param status The current [CheckboxStatus] of the checkbox.
 * @param onCheckboxClicked A lambda that is invoked when the user clicks the checkbox.
 * @param modifier The [Modifier] to be applied to the checkbox.
 * @param interactionSource Optional [MutableInteractionSource] used to observe interaction states
 *  like hover and press to drive visual feedback.
 * @param enabled A boolean that controls the enabled state of the checkbox.
 */
@Composable
public fun LemonadeUi.Checkbox(
    status: CheckboxStatus,
    onCheckboxClicked: () -> Unit,
    modifier: Modifier = Modifier,
    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
    enabled: Boolean = true,
) {
    CoreCheckbox(
        status = status,
        onCheckboxClicked = onCheckboxClicked,
        enabled = enabled,
        modifier = modifier,
    )
}

@Composable
private fun CoreCheckbox(
    status: CheckboxStatus,
    onCheckboxClicked: () -> Unit,
    enabled: Boolean,
    modifier: Modifier = Modifier,
    checkboxPlatformProps: CheckboxPlatformProps = platformCheckboxPropertiesProperties,
    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
) {
    val platformProps = platformCheckboxPropertiesProperties
    val isHovering by interactionSource.collectIsHoveredAsState()
    val isFocused by interactionSource.collectIsFocusedAsState()

    val isChecked = remember(status) {
        status in setOf(CheckboxStatus.Checked, CheckboxStatus.Indeterminate)
    }

    val animatedBorderColor by animateColorAsState(
        targetValue = when {
            !enabled && status == CheckboxStatus.Unchecked -> LocalColors.current.border.borderNeutralMedium
            enabled && status == CheckboxStatus.Unchecked -> LocalColors.current.border.borderNeutralHigh
            else -> LocalColors.current.border.borderNeutralHigh.copy(
                alpha = LocalOpacities.current.base.opacity0,
            )
        },
    )
    val animatedBackgroundColor by animateColorAsState(
        targetValue = when {
            !enabled -> LocalColors.current.background.bgElevatedHigh
            !isChecked && isHovering -> LocalColors.current.interaction.bgSubtleInteractive
            !isChecked -> LocalColors.current.background.bgDefault
            isChecked && isHovering -> LocalColors.current.interaction.bgBrandHighInteractive
            isHovering -> LocalColors.current.content.contentPrimaryInverse
            else -> LocalColors.current.background.bgBrandHigh
        },
    )
    val animatedIconTint by animateColorAsState(
        targetValue = if (enabled) {
            LocalColors.current.content.contentPrimaryInverse
        } else {
            LocalColors.current.content.contentTertiary
        },
    )

    Box(
        contentAlignment = Alignment.Center,
        modifier = modifier
            .focusable(
                enabled = enabled,
                interactionSource = interactionSource,
            )
            .clip(shape = LocalShapes.current.radius150)
            .clickable(
                role = Role.Checkbox,
                enabled = enabled,
                onClick = onCheckboxClicked,
                interactionSource = interactionSource,
                indication = null,
            )
            .then(
                other = if (isFocused && platformProps.focusVisible) {
                    Modifier
                        .border(
                            width = LocalBorderWidths.current.state.focusRing,
                            shape = LocalShapes.current.radius150,
                            color = LocalColors.current.border.borderSelected,
                        )
                        .padding(all = LocalBorderWidths.current.state.focusRing + 1.dp)
                        .clip(shape = LocalShapes.current.radius150)
                } else {
                    Modifier
                },
            )
            .border(
                width = LocalBorderWidths.current.base.border50,
                color = animatedBorderColor,
                shape = LocalShapes.current.radius150,
            )
            .background(color = animatedBackgroundColor)
            .requiredSize(size = platformProps.checkboxSize),
    ) {

        AnimatedContent(
            targetState = status.icon,
            transitionSpec = { fadeIn() + scaleIn() togetherWith fadeOut() + scaleOut() },
            content = { icon ->
                if (icon != null) {
                    LemonadeUi.Icon(
                        icon = icon,
                        contentDescription = status.name,
                        size = LemonadeAssetSize.Medium,
                        tint = animatedIconTint,
                    )
                } else {
                    Spacer(modifier = Modifier.fillMaxSize())
                }
            },
        )
    }
}


private val CheckboxStatus.icon: LemonadeIcons?
    get() {
        return when (this) {
            CheckboxStatus.Checked -> LemonadeIcons.CheckSmall
            CheckboxStatus.Unchecked -> null
            CheckboxStatus.Indeterminate -> LemonadeIcons.Minus
        }
    }

@Stable
internal data class CheckboxPlatformProps(
    val checkboxSize: Dp,
    val labelStyle: LemonadeTextStyle,
    val supportTextStyle: LemonadeTextStyle,
    val focusVisible: Boolean,
)

internal expect val platformCheckboxPropertiesProperties: CheckboxPlatformProps

@Composable
internal fun defaultPlatformCheckboxProps(): CheckboxPlatformProps {
    return CheckboxPlatformProps(
        checkboxSize = 22.dp,
        labelStyle = LocalTypographies.current.bodyMediumMedium,
        supportTextStyle = LocalTypographies.current.bodySmallRegular,
        focusVisible = false,
    )
}

@LemonadePreview
@Composable
internal fun LemonadeCheckboxPreview() {
    Column(
        modifier = Modifier.background(color = Color.White),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(
            space = 24.dp,
            alignment = Alignment.CenterVertically,
        )
    ) {
        CheckboxStatus.entries.forEach { status ->
            LemonadeUi.Checkbox(
                status = status,
                onCheckboxClicked = { /* Nothing */ },
                enabled = true,
            )
            LemonadeUi.Checkbox(
                status = status,
                onCheckboxClicked = { /* Nothing */ },
                enabled = false,
            )
        }
    }
}

@LemonadePreview
@Composable
internal fun LemonadeLabeledCheckboxPreview() {
    Column(
        modifier = Modifier.background(color = Color.White),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(
            space = 24.dp,
            alignment = Alignment.CenterVertically,
        )
    ) {
        CheckboxStatus.entries.forEach { status ->
            LemonadeUi.Checkbox(
                label = "Label",
                supportText = "Support text",
                status = status,
                onCheckboxClicked = { /* Nothing */ },
                enabled = true,
            )
            LemonadeUi.Checkbox(
                label = "Label",
                supportText = "Support text",
                status = status,
                onCheckboxClicked = { /* Nothing */ },
                enabled = false,
            )
        }
    }
}